File: 297. Serialize and Deserialize Binary Tree/solution.js
/** Definition for a binary tree node. */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function (root) {
    let data = [];
    function dfs(node) {
        if (node) {
            data.push(node.val);
            dfs(node.left);
            dfs(node.right);
        } else {
            // we also keep track of null nodes
            data.push('x');
        }
    }
    dfs(root);
    return data.join(' ');
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function (data) {
    function generateTree(tokens) {
        // we will pop to reduce the tokens array and get the next element
        let val = tokens.pop();
        if (val === 'x') {
            return null;
        } else {
            let node = new TreeNode(Number(val));
            node.left = generateTree(tokens);
            node.right = generateTree(tokens);
            return node;
        }
    }
    // we will reverse the string to make it easier to pop
    let tokens = data.split(' ').reverse();
    return generateTree(tokens);
};

File: 297. Serialize and Deserialize Binary Tree/solution.py
from typing import *

class TreeNode:
    """Definition for a binary tree node."""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Codec:
    def serialize(self, root: TreeNode) -> str:
        """Encodes a tree to a single string."""
        tokens = []
        def dfs(node):
            if not node:
                tokens.append('x')
            else:
                tokens.append(str(node.val))
                dfs(node.left)
                dfs(node.right)
        dfs(root)
        return ' '.join(tokens)

    def deserialize(self, data: str) -> TreeNode:
        """Decodes your encoded data to tree."""
        def genTree(tokens):
            val = next(tokens)
            if val == 'x':
                return None
            else:
                node = TreeNode(int(val))
                node.left = genTree(tokens)
                node.right = genTree(tokens)
                return node
        tokens = iter(data.split(' '))
        return genTree(tokens)